src/module/apps/dialogs/ShadowrunItemDialog.ts

```diff
@@ -1,8 +1,44 @@
-import { SR5Item } from '../../item/SR5Item';
-import { Helpers } from '../../helpers';
+import {SR5Item} from '../../item/SR5Item';
+import {Helpers} from '../../helpers';
+import {LENGTH_UNIT, SR} from "../../constants";
+import {SR5Actor} from "../../actor/SR5Actor";
+import FireModeData = Shadowrun.FireModeData;
+import RangesTemplateData = Shadowrun.RangesTemplateData;
+import RangeData = Shadowrun.RangeData;
 
+type ItemDialogData = {
+    dialogData: DialogData | undefined,
+    getActionTestData: Function | undefined,
+    itemHasNoDialog: boolean
+};
+
+export type RangedWeaponActionTestData = {
+    environmental: {
+        range?: number
+    },
+    fireMode: FireModeData,
+}
+
+export type SpellActionTestData = {
+    force: number,
+    reckless: boolean,
+}
+
+export type ComplexFormTestData = {
+    level: number
+}
+
+export type ActionTestData = {
+    rangedWeapon?: RangedWeaponActionTestData,
+    spell?: SpellActionTestData,
+    complexForm?: ComplexFormTestData,
+    targetId?: string
+}
+
+
+// TODO: Why extend dialog when internal dialog structures aren't used? Could resolve the whole flag / data flow issue.
```

Comment:
I think I was just extending Dialog because it seemed like the right thing to do without really thinking it through. If you feel like changing that I have no issue with it.

Reply:
I completly redid this at the current branch head with a custom FormDialog extension solving most of the promise logic / data flow inconsistencies. :)